text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1176 @@
+  LogEmitter methods of corresponding providers. This may not be the right
+  approach for languages where ABI stability is part of our guarantees since it
+  likely breaks the ABI.
+
+- Add a method overload that allows passing 3 parameters (instrumentation
+  library name, version and schema url) to obtain a Tracer/Meter/LogEmitter.
+  This is likely the preferred approach for languages where method overloads are
+  possible.
+
+- If neither of the above 2 approaches are possible to do in non-breaking manner
+  then the API may introduce a `SetSchema(schema_url)` method to
+  Tracer/Meter/LogEmitter instance. The method MUST be called only once and
+  MUST be called before any telemetry is emitted using the instance.
+
+There may be other ways to modify the API to allow the association with a Schema
+URL. Language maintainers SHOULD choose the idiomatic way for their language.
+
+The effect of associating a Schema URL with a Tracer/Meter/LogEmitter SHOULD be
+that the schema_url in the InstrumentationLibrarySpans,
+InstrumentationLibraryMetrics, InstrumentationLibraryLogs message for all the
+telemetry emitted with the associated Tracer/Meter/LogEmitter is populated with
+the provided Schema URL value.
+
+If the Tracer/Meter/LogEmitter is not associated with a Schema URL then the
+exporter MUST leave the schema_url field in OTLP messages unset, in which case
+the application-wide Schema URL [will apply](#application-wide-schema-url).
+
+Open Question: how to make it easy for instrumentation libraries to refer to a
+particular OpenTelemetry schema version and also make sure any semantic
+convention helpers the library uses (e.g. constants that define the semantic
+conventions) match exactly that same schema version? One possible solution is to
+introduce helper packages per schema version that the libraries can use, e.g.
+constants that define the semantic conventions and the corresponding schema
+version url. This should be likely the topic for a follow-up OTEP.
+
+### Application-wide Schema URL
+
+The SDK interface MUST provide a way for the user to optionally set an
+application-wide Schema URL. This Schema URL will be populated in all
+ResourceSpans, ResourceMetrics and ResourceLogs messages emitted by OTLP
+Exporter.
+
+If the user does not set an application-wide Schema URL then the current Schema
+URL MUST be populated by OTLP Exporter in the messages, where "current" means
+the version of OpenTelemetry Schema against which the SDK is coded.
+
+Note that if there is a schema url associated with instrumentation library it
+overrides the application-wide schema url as described [here](#otlp-changes).
+
+## OpenTelemetry Schema
+
+OpenTelemetry publishes it own schema at
+`https://opentelemetry.io/schemas/<version>`. The version number of the schema
+is the same as the specification version number which publishes the schema.
+Every time a new specification version is released a corresponding new schema
```

Comment:
It may be useful to require even stronger coupling. Since both schemas and semantic conventions are machine-readable, there could be a simple validation script that makes sure that the new version of the schema accurately reflects the changes to the spec, e.g. if an attribute was removed / renamed from the spec, something about that has to be present in the schema and we should block the release otherwise.

Reply:
Yes, validation is a good idea.